rust/lsp-lib/src/lsp_plugin.rs

```diff
@@ -0,0 +1,470 @@
+
+    let char_offset: usize = view.get_line(line_num)?[0..(offset - line_offset)]
+        .chars()
+        .map(char::len_utf16)
+        .sum();
+
+    Ok(Position {
+        line: line_num as u64,
+        character: char_offset as u64,
+    })
+}
+
+/// Get contents changes of a document modeled according to Language Server Protocol
+/// given the RopeDelta
+fn get_document_content_changes<C: Cache>(
+    delta: Option<&RopeDelta>,
+    view: &mut View<C>,
+) -> Result<Vec<TextDocumentContentChangeEvent>, PluginLibError> {
+    if let Some(delta) = delta {
+        if let Some(node) = delta.as_simple_insert() {
+            let (interval, _) = delta.summary();
+            let text = String::from(node);
+
+            let (start, end) = interval.start_end();
+            let text_document_content_change_event = TextDocumentContentChangeEvent {
+                range: Some(Range {
+                    start: get_position_of_offset(view, start)?,
+                    end: get_position_of_offset(view, end)?,
+                }),
+                range_length: Some((end - start) as u64),
+                text,
+            };
+
+            Ok(vec![text_document_content_change_event])
+        }
+        // Or a simple delete
+        else if delta.is_simple_delete() {
+            let (interval, _) = delta.summary();
+
+            let (start, end) = interval.start_end();
+
+            // Hack around sending VSCode Style Positions to Language Server.
+            // See this issue to understand: https://github.com/Microsoft/vscode/issues/23173
+            let mut end_position = get_position_of_offset(view, end)?;
+
+            if end_position.character == 0 {
+                let mut ep = get_position_of_offset(view, end - 1)?;
+                ep.character += 1;
+                end_position = ep;
+            }
+
+            let text_document_content_change_event = TextDocumentContentChangeEvent {
+                range: Some(Range {
+                    start: get_position_of_offset(view, start)?,
+                    end: end_position,
+                }),
+                range_length: Some((end - start) as u64),
+                text: String::new(),
+            };
+
+            Ok(vec![text_document_content_change_event])
+        }
+        // Send the whole document again if it is not a trivial edit
+        else {
+            let text_document_content_change_event = TextDocumentContentChangeEvent {
+                range: None,
+                range_length: None,
+                text: view.get_document()?,
+            };
+
+            Ok(vec![text_document_content_change_event])
+        }
+    } else {
+        let text_document_content_change_event = TextDocumentContentChangeEvent {
+            range: None,
+            range_length: None,
+            text: view.get_document()?,
+        };
+
+        Ok(vec![text_document_content_change_event])
+    }
+}
+
+pub fn get_workspace_root_uri(
```

Comment:
it sure feels like we should be able to reuse code from the earlier fn.

Reply:
The earlier function was unneeded after a refactor, so I removed it in favour of this. I forgot to remove it earlier.